src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    /// <summary>
+    /// Caches tokens on a per-line basis for documents based on semantic version.
+    /// </summary>
+    /// <remarks>
+    /// The cache makes the assumption that the tokens passed in are complete for the given range.
+    /// </remarks>
+    internal sealed class SemanticTokensCache
+    {
+        private const int MaxSemanticVersionPerDoc = 4;
+
+        // Nested cache mapping (URI -> (semanticVersion -> (line #s -> tokens on line)))
+        private readonly MemoryCache<DocumentUri, MemoryCache<VersionStamp, Dictionary<int, List<int>>>> _cache = new();
```

Comment:
This certainly seems like a big hammer at first glance, but now that I think of it caching the values by line seems more maintainable than having disparate ranges from different requests that we try to join by the ending points.

What happens if the range is halfway through the line and therefore the tokens are incomplete?

I'd vote to explicitly supply a limit, even if we're happy to use the MemoryCache default.

Reply:
> What happens if the range is halfway through the line and therefore the tokens are incomplete?

In practice from testing, the LSP client only sends us range requests for complete lines. Although you bring up a good point, we should be defensive here and probably just not utilize the cache if the client sends us a halfway-range. I'll add a case for that 👍

Reply:
> we should be defensive here and probably just not utilize the cache if the client sends us a halfway-range

Yeah, they could easily change that behavior or a new client might behave differently.